iree/compiler/Conversion/LinalgToLLVM/PadLinalgWorkgroupTiles.cpp

```diff
@@ -0,0 +1,265 @@
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.createBlock(&region, region.end(), blockArgTypes);
+  rewriter.create<linalg::YieldOp>(loc, padding);
+  return padTensorOp;
+}
+
+// Returns padding for dim to next integer multiple of the vector size.
+static std::pair<int, int> getVectorPaddingSize(int dim, int vecSize) {
+  if (dim < vecSize) {
+    int size = vecSize;
+    int padValue = size - dim;
+    return {vecSize, padValue};
+  } else {
+    int size = ceil((float)dim / (float)vecSize) * vecSize;
+    int padValue = size - dim;
+    return {size, padValue};
+  }
+}
+
+// Returns padding for dim to next integer multiple of the workgroup size.
+// Note: KernelDispatch gurantess workgroup size is the largest integer
+// multiple of the vector size.
+std::pair<int, int> getWorkgroupPaddedTileSize(int dim, int workgoupSize,
+                                               int vecSize) {
+  if (dim > workgoupSize) {
+    int size = ceil((float)dim / (float)workgoupSize) * workgoupSize;
+    int padValue = size - dim;
+    return {workgoupSize, padValue};
+  } else {
+    return getVectorPaddingSize(dim, vecSize);
+  }
+}
+
+// Clones all attributes from src to dst Operation.
+static void cloneAttributes(mlir::Operation *src, mlir::Operation *dst) {
+  for (auto attr : src->getAttrs()) {
+    dst->setAttr(attr.first, attr.second);
+  }
+}
+
+// Returns static full-shape sizes of operand.
+ArrayRef<int64_t> getFullSize(Value operand) {
+  auto defOp = operand.getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+  if (!defOp) return {};
+  TensorType shape = defOp.source()
+                         .getType()
+                         .cast<IREE::Flow::DispatchTensorType>()
+                         .asTensorType();
+  if (!shape.hasStaticShape()) return {};
+  return shape.getShape();
+}
+
+// Creates linalg.matmul with operands padded to the next integer multiple of
+// the workgroup size.
+class MatmulWorkgroupTilesPadding : public OpRewritePattern<linalg::MatmulOp> {
+ public:
+  using OpRewritePattern<linalg::MatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = matmulOp.getLoc();
+    auto lhs = matmulOp.getInput(0);
+    auto rhs = matmulOp.getInput(1);
+    auto result = matmulOp.getOutput(0);
+
+    if (lhs.getDefiningOp<linalg::PadTensorOp>() ||
+        rhs.getDefiningOp<linalg::PadTensorOp>())
+      return failure();
+
+    auto lhsFullSize = getFullSize(lhs);
+    auto rhsFullSize = getFullSize(rhs);
+
+    if (lhsFullSize.empty() || rhsFullSize.empty()) return failure();
+
+    int problemSizeM = lhsFullSize[0];
+    int problemSizeN = rhsFullSize[1];
+    int problemSizeK = lhsFullSize[1];
+
+    auto workgroupTileSizes =
+        getTileSizes<TilingLevel::WorkGroupTiles>(matmulOp);
+    auto vectorTileSizes = getTileSizes<TilingLevel::Level2Tiles>(matmulOp);
```

Comment:
This is going away. PR #5842 removes this function. Instead it should be obtained from the attribute that is added by "materializeStatic..." pass. Do you want to rebase on top of that?

Reply:
Actually strike that. Its easy to rebase on this for me. So submit this after other comments are addressed

Reply:
SG! I don't see any comments :)

Reply:
Please add a todo for me to fix after PR #5842 lands.